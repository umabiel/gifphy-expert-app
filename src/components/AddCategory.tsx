import React, {
  ChangeEvent,
  Dispatch,
  FormEvent,
  Fragment,
  SetStateAction,
  useState,
} from "react";

import PropTypes from "prop-types";

interface IProps {
  setCategories: Dispatch<SetStateAction<string[]>>;
}

export const AddCategory = ({ setCategories }: IProps) => {
  const [inputValue, setInputValue] = useState("");
  const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    setInputValue(e.target.value);
  };

  const handleSubmit = (e: FormEvent) => {
    e.preventDefault();
    if (inputValue.trim().length > 2) {
      setCategories((animes) => [inputValue, ...animes]);
      setInputValue("");
    }
  };
  return (
    <Fragment>
      <form onSubmit={handleSubmit}>
        <input type="text" value={inputValue} onChange={handleInputChange} />
      </form>
    </Fragment>
  );
};

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired,
};
