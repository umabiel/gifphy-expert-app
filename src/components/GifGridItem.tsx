import { IGifs } from "../core/oop/interfaces/IGifs";

export const GifGridItem = ({ id, title, url }: IGifs) => {
  return (
    <div className="card">
      <p>{title}</p>
      <img src={url} alt={title} />
    </div>
  );
};
