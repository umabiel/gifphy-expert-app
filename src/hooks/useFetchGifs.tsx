import { useEffect, useState } from "react";
import { getGifs } from "../helpers/getGifs";

interface IGifs {
  id: string;
  title: string;
  url: string;
}

export const useFetchGifs = (category: string) => {
  const initialState = {
    data: [] as IGifs[], // Shameless Hack u_u
    loading: true,
  };
  const [state, setState] = useState(initialState);

  useEffect(() => {
    getGifs(category).then((res) => {
      setState({
        data: res,
        loading: false,
      });
    });
  }, [category]);

  return state;
};
