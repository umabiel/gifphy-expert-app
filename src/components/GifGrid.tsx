import React, { Fragment } from "react";
import { IGifs } from "../core/oop/interfaces/IGifs";
import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifGridItem } from "./GifGridItem";
import PropTypes from "prop-types";

interface IProps {
  category: string;
}

export const GifGrid = ({ category = "One Punch" }: IProps) => {
  // const initState: IGifs[] = [];
  // const [images, setImages] = useState(initState);
  const { data: images, loading } = useFetchGifs(category);

  return (
    <Fragment>
      <h3>{category}</h3>
      {loading && <p>Loading...</p>}
      <div className="card-grid">
        {images.map((x: IGifs) => (
          <GifGridItem key={x.id} {...x} />
        ))}
      </div>
    </Fragment>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired,
};
