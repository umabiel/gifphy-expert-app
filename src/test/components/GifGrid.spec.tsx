import { shallow } from "enzyme";
import React from "react";
import { GifGrid } from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";

jest.mock("../../hooks/useFetchGifs");

describe("Pruebas en <GifGrid />", () => {
  const category = "";
  test("Debe renderizar el componente <GifGrid />", () => {
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true,
    });
    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe de mostrar <GifGridItem /> cuando se cargan imagenes con el Hooks useFetchGifs", () => {
    const gifs = [
      {
        id: "ABC",
        url: "http://localhost/abc.png",
        title: "Titutlo ABC",
      },
    ];
    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false,
    });
    const wrapper = shallow(<GifGrid category={category} />);
    // expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("p").exists()).toBe(false);
    expect(wrapper.find("GifGridItem").length).toEqual(gifs.length);
  });
});
