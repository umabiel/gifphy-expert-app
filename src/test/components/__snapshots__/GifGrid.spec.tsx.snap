// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGrid /> Debe renderizar el componente <GifGrid /> 1`] = `
<Fragment>
  <h3 />
  <p>
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
